torch/utils/data/distributed.py

```diff
@@ -65,3 +65,112 @@ def __len__(self):
 
     def set_epoch(self, epoch):
         self.epoch = epoch
+
+class ChunkDataReader(object):
```

Comment:
We really shouldn't have `ChunkDataReader`s. This class is the same thing as an iterable dataset.

Reply:
This implementation hides (iterable) dataset implementation from the user along with coordinating data loading amongst workers (including multiple nodes). This PRimplements `ChunkDataset` so that future users forget about it. All the user will have to do is implementing a data format specific function (`ChunkDataReader`) that can handle `chunks`, wherever chunks are in their scenario. Current `IterableDataset` works perfectly for non-distributed scenarios, but requires lots of boiler plate to accomplish the distributed version. That is the extension we are proposing: makes `IterableDataset` easier/more powerful for distributed training with less coding.

In the scenario that you are describing (removing chunk data reader) and using iterable dataset, how do you propose each worker would read, load and shuffle only the slices of dataset they strictly need?

Reply:
You said that `IterableDataset` would be harder to deal with, but I don't fully understand why? What's so special about this class? How do you handle replication of the state in a distributed fashion that you couldn't reproduce with a regular dataset?

Reply:
Oh sorry, I didn't notice that it's actually stateless. In that case why couldn't it be a regular map-style dataset?

Reply:
The original design changed a bit - or at least will change as soon as https://github.com/pytorch/pytorch/pull/28841 lands.

`ChunkDataReader` class will be replaced by a new method on `ChunkDataset(IterableDataset)` class. This method will be responsible for fetching data for a specific chunk index.

`DistributedChunkSampler` is also being eliminated after https://github.com/pytorch/pytorch/pull/28841.

There are some subtle changes that can be found in the discussion below, but the core is the same: creating a `ChunkDataset` (which is `IterableDataset`) that can be used in distributed training using a distributed sampler and which also saves users from boiler plate code to handle dataset (with unknown size) reading in each worker.